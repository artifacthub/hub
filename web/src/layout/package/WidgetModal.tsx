import React, { useEffect, useState } from 'react';
import { FiMoon, FiSun } from 'react-icons/fi';
import { useHistory } from 'react-router-dom';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { docco } from 'react-syntax-highlighter/dist/cjs/styles/hljs';

import { SearchFiltersURL } from '../../types';
import ButtonCopyToClipboard from '../common/ButtonCopyToClipboard';
import Modal from '../common/Modal';
import styles from './WidgetModal.module.css';

interface Props {
  packageId: string;
  packageName: string;
  packageDescription: string;
  visibleWidget: boolean;
  searchUrlReferer?: SearchFiltersURL;
  fromStarredPage?: boolean;
  setOpenStatus: React.Dispatch<React.SetStateAction<boolean>>;
}

interface WidgetTheme {
  name: string;
  icon: JSX.Element;
}

const DEFAULT_THEME = 'light';
const THEMES: WidgetTheme[] = [
  {
    name: DEFAULT_THEME,
    icon: <FiSun />,
  },
  { name: 'dark', icon: <FiMoon /> },
];

const WidgetModal = (props: Props) => {
  const history = useHistory();
  const [theme, setTheme] = useState<string>(DEFAULT_THEME);
  const [header, setHeader] = useState<boolean>(true);
  const [responsive, setRepsonsive] = useState<boolean>(false);

  const buildWidgetCode = (): string => {
    const url = `${window.location.origin}${window.location.pathname}`;
    const code = `<div class="artifacthub-widget" data-url="${url}" data-theme="${theme}" data-header="${
      !header ? 'false' : 'true'
    }" data-responsive="${responsive ? 'true' : 'false'}"><blockquote><p lang="en" dir="ltr"><b>${
      props.packageName
    }</b>${
      props.packageDescription ? `: ${props.packageDescription}` : ''
    }</p>&mdash; Open in <a href="${url}">Artifact Hub</a></blockquote></div><script async src="${
      window.location.origin
    }/artifacthub-widget.js"></script>`;

    return code;
  };

  const [widgetCode, setWidgetCode] = useState<string>(buildWidgetCode());

  const resetValues = () => {
    setTheme(DEFAULT_THEME);
    setHeader(true);
    setRepsonsive(false);
    setWidgetCode(buildWidgetCode());
  };

  const onCloseModal = () => {
    props.setOpenStatus(false);
    resetValues();
    history.replace({
      search: '',
      state: { searchUrlReferer: props.searchUrlReferer, fromStarredPage: props.fromStarredPage },
    });
  };

  useEffect(() => {
    setWidgetCode(buildWidgetCode());
    /* eslint-disable react-hooks/exhaustive-deps */
  }, [theme, header, responsive, props.packageId, props.visibleWidget]);
  /* eslint-enable react-hooks/exhaustive-deps */

  useEffect(() => {
    if (props.visibleWidget) {
      history.replace({
        search: '?modal=widget',
        state: { searchUrlReferer: props.searchUrlReferer, fromStarredPage: props.fromStarredPage },
      });
    }
  }, [props.visibleWidget]); /* eslint-disable-line react-hooks/exhaustive-deps */

  return (
    <>
      {props.visibleWidget && (
        <Modal
          modalDialogClassName={styles.modalDialog}
          header={<div className={`h3 m-2 flex-grow-1 ${styles.title}`}>Widget</div>}
          onClose={onCloseModal}
          open={props.visibleWidget}
        >
          <div className="w-100 position-relative">
            <label className={`font-weight-bold ${styles.label}`} htmlFor="theme">
              Theme
            </label>
            <div className="d-flex flex-row mb-3">
              {THEMES.map((themeOpt: WidgetTheme) => {
                return (
                  <div className="custom-control custom-radio mr-4" key={`radio_theme_${themeOpt.name}`}>
                    <input
                      className="custom-control-input"
                      type="radio"
                      name="theme"
                      id={themeOpt.name}
                      value={themeOpt.name}
                      checked={theme === themeOpt.name}
                      required
                      readOnly
                    />
                    <label
                      className="text-capitalize custom-control-label"
                      htmlFor={themeOpt.name}
                      onClick={() => {
                        setTheme(themeOpt.name);
                      }}
                    >
                      <div className="d-flex flex-row align-items-center">
                        {themeOpt.icon}
                        <span className="ml-1">{themeOpt.name}</span>
                      </div>
                    </label>
                  </div>
                );
              })}
            </div>

            <div className="mt-4 mb-3">
              <div className="custom-control custom-switch pl-0">
                <input
                  id="header"
                  type="checkbox"
                  className="custom-control-input"
                  value="true"
                  onChange={() => setHeader(!header)}
                  checked={header}
                />
                <label
                  htmlFor="header"
                  className={`custom-control-label font-weight-bold ${styles.label} ${styles.customControlRightLabel}`}
                >
                  Header
                </label>
              </div>

              <small className="form-text text-muted mt-2">
                Displays Artifact Hub header at the top of the widget.
              </small>
            </div>

            <div className="mt-4 mb-3">
              <div className="custom-control custom-switch pl-0">
                <input
                  id="responsive"
                  type="checkbox"
                  className="custom-control-input"
                  value="true"
                  onChange={() => setRepsonsive(!responsive)}
                  checked={responsive}
                />
                <label
                  htmlFor="responsive"
                  className={`custom-control-label font-weight-bold ${styles.label} ${styles.customControlRightLabel}`}
                >
                  Responsive
                </label>
              </div>

              <small className="form-text text-muted mt-2">
                The widget will try to use the width available on the parent container (between 350px and 650px).
              </small>
            </div>

            <div className="mt-3 mb-2">
              <label className={`font-weight-bold ${styles.label}`}>Code</label>

              <div className={`flex-grow-1 mr-3 user-select-none ${styles.blockWrapper}`}>
                <SyntaxHighlighter
                  language="html"
                  style={docco}
                  customStyle={{
                    backgroundColor: 'var(--color-1-10)',
                  }}
                >
                  {widgetCode}
                </SyntaxHighlighter>
              </div>

              <ButtonCopyToClipboard
                text={widgetCode}
                tooltipClassName={`bs-tooltip-right ${styles.copyBtnTooltip}`}
                arrowClassName={styles.copyBtnArrow}
                visibleBtnText
                contentBtn="Copy code to clipboard"
                className={`btn-secondary ${styles.copyBtn}`}
              />
            </div>
          </div>
        </Modal>
      )}
    </>
  );
};

export default WidgetModal;
