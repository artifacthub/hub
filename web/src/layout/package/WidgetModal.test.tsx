import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter as Router } from 'react-router-dom';

import WidgetModal from './WidgetModal';

jest.mock('react-color', () => ({
  SketchPicker: () => <>sketch</>,
}));

const setOpenStatusMock = jest.fn();

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as {}),
  useNavigate: () => mockUseNavigate,
}));

jest.mock('react-syntax-highlighter', () => (props: any) => <div>{props.children}</div>);

const defaultProps = {
  packageId: 'id',
  packageName: 'pkg',
  packageDescription: 'this is the description',
  visibleWidget: true,
  setOpenStatus: setOpenStatusMock,
};

describe('WidgetModal', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <Router>
        <WidgetModal {...defaultProps} />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      render(
        <Router>
          <WidgetModal {...defaultProps} />
        </Router>
      );

      expect(screen.getByRole('dialog')).toBeInTheDocument();
      expect(screen.getByText('Theme')).toBeInTheDocument();
      expect(screen.getByText('light')).toBeInTheDocument();
      expect(screen.getByText('dark')).toBeInTheDocument();
      expect(screen.getByRole('radio', { name: 'light' })).toBeChecked();
      expect(screen.getByRole('radio', { name: 'dark' })).not.toBeChecked();
      expect(screen.getByText('Responsive')).toBeInTheDocument();
      expect(screen.getByRole('switch', { name: 'Responsive' })).not.toBeChecked();
      expect(
        screen.getByText(
          'The widget will try to use the width available on the parent container (between 350px and 650px).'
        )
      ).toBeInTheDocument();
      expect(screen.getByText('Stars')).toBeInTheDocument();
      expect(screen.getByRole('switch', { name: 'Stars' })).toBeChecked();
      expect(screen.getByText('Display number of stars given to the package.')).toBeInTheDocument();
      expect(screen.getByText('Code')).toBeInTheDocument();
      expect(screen.getByTestId('block-content')).toHaveTextContent(
        '<div class="artifacthub-widget" data-url="http://localhost/" data-theme="light" data-header="true" data-stars="true" data-responsive="false"><blockquote><p lang="en" dir="ltr"><b>pkg</b>: this is the description</p>&mdash; Open in <a href="http://localhost/">null</a></blockquote></div><script async src="http://localhost/artifacthub-widget.js"></script>'
      );
    });

    it('when not white label', () => {
      render(
        <Router>
          <meta name="artifacthub:siteName" content="artifact hub" />
          <WidgetModal {...defaultProps} />
        </Router>
      );

      expect(screen.getByText('Header')).toBeInTheDocument();
      expect(screen.getByRole('switch', { name: 'Header' })).toBeChecked();
      expect(screen.getByText('Display Artifact Hub header at the top of the widget.')).toBeInTheDocument();
    });

    it('updates block content to change different options', async () => {
      render(
        <Router>
          <WidgetModal {...defaultProps} />
        </Router>
      );

      expect(screen.getByTestId('block-content')).toHaveTextContent(
        '<div class="artifacthub-widget" data-url="http://localhost/" data-theme="light" data-header="true" data-stars="true" data-responsive="false"><blockquote><p lang="en" dir="ltr"><b>pkg</b>: this is the description</p>&mdash; Open in <a href="http://localhost/">null</a></blockquote></div><script async src="http://localhost/artifacthub-widget.js"></script>'
      );

      await userEvent.click(screen.getByText('dark'));
      await userEvent.click(screen.getByText('Responsive'));

      expect(screen.getByTestId('block-content')).toHaveTextContent(
        '<div class="artifacthub-widget" data-url="http://localhost/" data-theme="dark" data-header="true" data-stars="true" data-responsive="true"><blockquote><p lang="en" dir="ltr"><b>pkg</b>: this is the description</p>&mdash; Open in <a href="http://localhost/">null</a></blockquote></div><script async src="http://localhost/artifacthub-widget.js"></script>'
      );
    });
  });
});
