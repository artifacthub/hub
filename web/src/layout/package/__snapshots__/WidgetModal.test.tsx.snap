// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetModal creates snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="modal-backdrop activeBackdrop"
      data-testid="modalBackdrop"
    />
    <div
      aria-modal="true"
      class="modal modal active d-block"
      role="dialog"
    >
      <div
        class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
      >
        <div
          class="modal-content border border-3 mx-auto content"
        >
          <div
            class="modal-header d-flex flex-row align-items-center header undefined"
          >
            <div
              class="h3 m-2 flex-grow-1 title"
            >
              Widget
            </div>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body p-4 h-100 d-flex flex-column"
          >
            <div
              class="w-100 position-relative"
            >
              <label
                class="form-label fw-bold label"
                for="theme"
              >
                Theme
              </label>
              <div
                class="d-flex flex-row mb-3"
              >
                <div
                  class="form-check me-4"
                >
                  <input
                    checked=""
                    class="form-check-input"
                    id="light"
                    name="theme"
                    required=""
                    type="radio"
                    value="light"
                  />
                  <label
                    class="form-label text-capitalize form-check-label"
                    for="light"
                  >
                    <div
                      class="d-flex flex-row align-items-center"
                    >
                      <svg
                        fill="none"
                        height="1em"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <circle
                          cx="12"
                          cy="12"
                          r="5"
                        />
                        <line
                          x1="12"
                          x2="12"
                          y1="1"
                          y2="3"
                        />
                        <line
                          x1="12"
                          x2="12"
                          y1="21"
                          y2="23"
                        />
                        <line
                          x1="4.22"
                          x2="5.64"
                          y1="4.22"
                          y2="5.64"
                        />
                        <line
                          x1="18.36"
                          x2="19.78"
                          y1="18.36"
                          y2="19.78"
                        />
                        <line
                          x1="1"
                          x2="3"
                          y1="12"
                          y2="12"
                        />
                        <line
                          x1="21"
                          x2="23"
                          y1="12"
                          y2="12"
                        />
                        <line
                          x1="4.22"
                          x2="5.64"
                          y1="19.78"
                          y2="18.36"
                        />
                        <line
                          x1="18.36"
                          x2="19.78"
                          y1="5.64"
                          y2="4.22"
                        />
                      </svg>
                      <span
                        class="ms-1"
                      >
                        light
                      </span>
                    </div>
                  </label>
                </div>
                <div
                  class="form-check me-4"
                >
                  <input
                    class="form-check-input"
                    id="dark"
                    name="theme"
                    required=""
                    type="radio"
                    value="dark"
                  />
                  <label
                    class="form-label text-capitalize form-check-label"
                    for="dark"
                  >
                    <div
                      class="d-flex flex-row align-items-center"
                    >
                      <svg
                        fill="none"
                        height="1em"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
                        />
                      </svg>
                      <span
                        class="ms-1"
                      >
                        dark
                      </span>
                    </div>
                  </label>
                </div>
              </div>
              <div
                class="mt-4 mb-3"
              >
                <div
                  class="form-check form-switch ps-0"
                >
                  <label
                    class="form-check-label fw-bold label"
                    for="header"
                  >
                    Header
                  </label>
                   
                  <input
                    checked=""
                    class="form-check-input position-absolute ms-2"
                    id="header"
                    role="switch"
                    type="checkbox"
                    value="true"
                  />
                </div>
                <div
                  class="form-text text-muted mt-2"
                >
                  Display Artifact Hub header at the top of the widget.
                </div>
              </div>
              <div
                class="mt-4 mb-3"
              >
                <div
                  class="form-check form-switch ps-0"
                >
                  <label
                    class="form-check-label fw-bold label"
                    for="stars"
                  >
                    Stars
                  </label>
                  <input
                    checked=""
                    class="form-check-input position-absolute ms-2"
                    id="stars"
                    role="switch"
                    type="checkbox"
                    value="true"
                  />
                </div>
                <div
                  class="form-text text-muted mt-2"
                >
                  Display number of stars given to the package.
                </div>
              </div>
              <div
                class="mt-4 mb-3"
              >
                <div
                  class="form-check form-switch ps-0"
                >
                  <label
                    class="form-check-label fw-bold label"
                    for="responsive"
                  >
                    Responsive
                  </label>
                  <input
                    class="form-check-input position-absolute ms-2"
                    id="responsive"
                    role="switch"
                    type="checkbox"
                    value="true"
                  />
                </div>
                <div
                  class="form-text text-muted mt-2"
                >
                  The widget will try to use the width available on the parent container (between 350px and 650px).
                </div>
              </div>
              <div
                class="mt-3 mb-2"
              >
                <label
                  class="form-label fw-bold label"
                >
                  Code
                </label>
                <div
                  class="flex-grow-1 me-3 user-select-none blockWrapper"
                  data-testid="block-content"
                >
                  <div>
                    &lt;div class="artifacthub-widget" data-url="http://localhost/" data-theme="light" data-header="true" data-stars="true" data-responsive="false"&gt;&lt;blockquote&gt;&lt;p lang="en" dir="ltr"&gt;&lt;b&gt;pkg&lt;/b&gt;: this is the description&lt;/p&gt;&mdash; Open in &lt;a href="http://localhost/"&gt;null&lt;/a&gt;&lt;/blockquote&gt;&lt;/div&gt;&lt;script async src="http://localhost/artifacthub-widget.js"&gt;&lt;/script&gt;
                  </div>
                </div>
                <div
                  class="position-relative undefined"
                >
                  <button
                    aria-label="Copy code to clipboard"
                    class="btn btn-sm btn-outline-secondary text-uppercase px-2 py-1"
                    type="button"
                  >
                    <div
                      aria-hidden="true"
                      class="d-flex flex-row align-items-center"
                    >
                      <div
                        class="me-2"
                      >
                        Copy code to clipboard
                      </div>
                      <svg
                        fill="none"
                        height="1em"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <rect
                          height="13"
                          rx="2"
                          ry="2"
                          width="13"
                          x="9"
                          y="9"
                        />
                        <path
                          d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
                        />
                      </svg>
                    </div>
                  </button>
                </div>
              </div>
            </div>
            <div>
              <div
                class="overflow-hidden alertWrapper"
                data-testid="alertWrapper"
              />
            </div>
          </div>
          <div
            class="modal-footer p-3 undefined"
          >
            <button
              aria-label="Close modal"
              class="btn btn-sm btn-outline-secondary text-uppercase"
              type="button"
            >
              <div
                class="d-flex flex-row align-items-center"
              >
                <svg
                  class="me-2"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
                <div>
                  Close
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
